Add seconds to DateTimePicker view

JamesMHenderson opened the pull request:
- Add seconds as an available view in DateTimePickerView
- Use DateTimePickerView type as the type for views and openTo
- Change typography variant depending on whether seconds are being displayed or not
- set number of grid elements depending on seconds view

<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [ x ] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/master/CONTRIBUTING.md#sending-a-pull-request).

dmtrKovalenko commented on lib/src/DateTimePicker/DateTimePicker.tsx:
```diff
@@ -14,15 +14,15 @@ import {
 export interface DateTimePickerViewsProps extends BaseDateTimePickerProps {
   /** Array of views to show */
-  views?: ('year' | 'date' | 'month' | 'hours' | 'minutes')[];
+  views?: DateTimePickerView[];
```
This was required to conrrectly display type in the documentation. It is now will be displaying as `DateTimePickerView`

dmtrKovalenko commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -42,6 +45,9 @@ export const DateTimePickerToolbar: React.FC<ToolbarComponentProps> = ({
   const theme = useTheme();
   const rtl = theme.direction === 'rtl';
 
+  const isSeconds = useMemo(() => arrayIncludes(views, 'seconds'), [views]);
```
```suggestion
  const hasSeconds = useMemo(() => arrayIncludes(views, 'seconds'), [views]);
```

dmtrKovalenko reviewed (changes requested):
This PR missing a test case.

dmtrKovalenko reviewed (changes requested):
Please fix all CI errors

JamesMHenderson commented:
@dmtrKovalenko the CI was failing from the branch, I branched off https://github.com/mui-org/material-ui-pickers/commit/32a3a1846a3f9762002c57b4b576954eb7685da4.

How would I go about fixing the CI errors as they were not introduced by these changes?

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -68,26 +74,39 @@ export const DateTimePickerToolbar: React.FC<ToolbarComponentProps> = ({
           >
             <ToolbarButton
-              variant="h3"
+              variant={timeTypographyText}
```
This could be an opportunity to fix the `<hx>` DOM output

JamesMHenderson replied:
What do you mean by this?

oliviertassinari commented on lib/src/__tests__/e2e/DateTimePickerRoot.test.tsx:
```diff
@@ -70,3 +70,33 @@ describe('e2e - DateTimePicker', () => {
+  let component: ReactWrapper<DateTimePickerProps>;
+  const onChangeMock = jest.fn();
+
+  beforeEach(() => {
```
Please avoid test preparation ceremony

```suggestion
```

oliviertassinari commented on lib/src/__tests__/e2e/DateTimePickerRoot.test.tsx:
```diff
@@ -70,3 +70,33 @@ describe('e2e - DateTimePicker', () => {
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    component = mount(
```
Please prefer testing-library for this

JamesMHenderson replied:
I've based my test on existing tests which are using enzyme, is it prefered that the tests use testing-library over being consistent with the existing tests?

oliviertassinari replied:
@JamesMHenderson the objective is to, step by step, be all in on testing-library. It requires all new tests to use it. This holds for all our codebase (core, pickers, etc). For instance, on the pickers side https://github.com/mui-org/material-ui-pickers/issues/1933.

oliviertassinari commented on lib/src/__tests__/e2e/DateTimePickerRoot.test.tsx:
```diff
@@ -70,3 +70,33 @@ describe('e2e - DateTimePicker', () => {
+        value={utilsToUse.date('2018-01-01T00:00:00.000Z')}
+        onChange={onChangeMock}
+        openTo="hours"
+        leftArrowIcon="keyboard_arrow_left"
+        rightArrowIcon="keyboard_arrow_right"
+        dateRangeIcon="date_range"
+        timeIcon="access_time"
```
This is not required and shouldn't work in the first place

```suggestion
```

oliviertassinari reviewed (changes requested):
Please rebase on the `next` branch

JamesMHenderson committed:
Add seconds to DateTimePicker view

- Add seconds as an available view in DateTimePickerView
- Use DateTimePickerView type as the type for views and openTo
- Change typography variant depending on whether seconds are being displayed or not
- set number of grid elements depending on seconds view

oliviertassinari commented on lib/src/__tests__/DateTimePickerTestingLib.test.tsx:
```diff
@@ -63,4 +63,41 @@ describe('<DateTimePicker />', () => {
+    );
+
+    await waitFor(() => screen.getByRole('dialog'));
+    expect(getByMuiTest('seconds').firstChild?.textContent).toBe('05');
```
```suggestion
    expect(getByMuiTest('seconds')).to.have.text('05');
```

JamesMHenderson replied:
I don't think this is valid

I get a type error with this change. `Property 'to' does not exist on type 'JestMatchersShape  , Matchers  , string | null>>'.`

oliviertassinari replied:
Oh right, It's probably because we miss

https://github.com/mui-org/material-ui/blob/d1287f1874dffeade600364b36727fa7878e9a75/test/utils/initMatchers.js#L16

oliviertassinari replied:
Ok, forget about it, we will have it resolved by merging the repository in the main one.

oliviertassinari commented on lib/src/__tests__/DateTimePickerTestingLib.test.tsx:
```diff
@@ -63,4 +63,41 @@ describe('<DateTimePicker />', () => {
+    fireEvent.click(getByMuiTest('seconds'));
+
+    expect(getByText(getByMuiTest('seconds'), '05')).toHaveClass(
+      'MuiPickersToolbarText-toolbarBtnSelected'
```
Oh nice, I'm adding this case to #1845.

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -117,25 +123,45 @@ export const DateTimePickerToolbar: React.FC<ToolbarComponentProps> = withDefaul
+                    className={classes.separator}
+                  />
+
+                  <ToolbarButton
```
The HTLM output is wrong:

<img width="316" alt="Capture d’écran 2020-07-08 à 12 19 08" src="https://user-images.githubusercontent.com/3165635/86907451-57384180-c115-11ea-9ff3-e19106d8b995.png">

1. We shouldn't have any headers in here, as far as I know. If we want to, the h1 > h2 > h3… order is broken
2. button > span > h4 nesting structure is wrong. I propose we remove the h4 element (we can use the component prop to compose behavior (button) and style (text).

JamesMHenderson replied:
I agree with you, but I'm not sure what the best approach would be for this. Would it be best to pick this sort of change up seperately from this pull request as it would be a significant refactor?

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -117,25 +123,45 @@ export const DateTimePickerToolbar: React.FC<ToolbarComponentProps> = withDefaul
+                    onClick={() => setOpenView('seconds')}
+                    selected={openView === 'seconds'}
+                    value={date ? utils.format(date, 'seconds') : '--'}
+                    typographyClassName={classes.timeTypography}
```
```suggestion
                    className={classes.timeTypography}
```

JamesMHenderson replied:
The Toolbar button has a prop typography class which is passed in to add the class to the typography component in the ToolbarButton, rather than adding the class to the Button directly.

oliviertassinari replied:
Yeah, there is an opportunity here to flatten the DOM structure '

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -117,25 +123,45 @@ export const DateTimePickerToolbar: React.FC<ToolbarComponentProps> = withDefaul
+                  <ToolbarButton
+                    tabIndex={-1}
+                    variant={timeTypographyText}
+                    data-mui-test="seconds"
```
Considering we want to move to `data-testid` in the whole codebase, I suggest we start here
```suggestion
                    data-testid="seconds"
```

cc @mui-org/core-team.

eglavin commented:
Hi fella any progress on this pull request? Anything i might be able to help with? Id love to implement this change into the app im working on, so if i can be of any help id be able to give some time!

oliviertassinari commented:
@eglavin We are moving the date picker components into the lab. We will archive this repository once done. Any effort will need to be rebased from the main repository.

oliviertassinari commented:
@JamesMHenderson Thanks for working on this problem. I'm closing as we have moved the component to https://github.com/mui-org/material-ui.
